### PR TITLE
Allow access expressions on all `BLangExpression`s

### LINKED ISSUE #30038: Access expressions should be allowed on all expressions
**Description:**
$title. ref: https://github.com/ballerina-platform/ballerina-lang/pull/30031#issuecomment-821768607

**Steps to reproduce:**
n/a

**Affected Versions:**
n/a

**OS, DB, other environment details and versions:**
n/a

**Related Issues (optional):**
- #29970
- #29453
- https://github.com/ballerina-platform/ballerina-lang/pull/30031

### PR DESCRIPTION
**Pull request by grainier**
## Purpose
$title.

Fixes #30038

## Approach
n/a

## Samples
n/a

## Remarks
 - https://github.com/ballerina-platform/ballerina-lang/pull/30031#issuecomment-821768607
 - https://github.com/ballerina-platform/ballerina-lang/pull/30039

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangExpression.java [hunk 1]
```diff
@@ -51,4 +52,6 @@
     public BType expectedType;
 
     public Map<BVarSymbol, NarrowedTypes> narrowedTypeInfo;
+
+    public BSymbol symbol;
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangExpression.java:56]: Wouldn't adding a symbol to `BLangExpression` be a bit misleading considering a symbol would only be associated in a few specific exprs such as var ref, function call etc? 🤔

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangExpression.java:56]: Yeah AFAICT there don't seem to be places where we used to access the symbol via the previous BLangAccessibleExpression. If we can add it to only the relevant expressions, shall we do so?

Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangExpression.java]: fixed: 89e2edd

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -2618,20 +2585,14 @@ public void visit(BLangIndexBasedAccess indexBasedAccessExpr) {
             return;
         }
 
-        boolean isStringValue = containerExpression.type != null && containerExpression.type.tag == TypeTags.STRING;
-
-        // Container expression must be a accessible expression.
-        if (!(containerExpression instanceof BLangAccessibleExpression) && !isStringValue) {
-            dlog.error(containerExpression.pos, DiagnosticErrorCode.EXPRESSION_DOES_NOT_SUPPORT_MEMBER_ACCESS,
```
[the remark this answers is not in the document]
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:2625]: fixed: 1057b81

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangValueExpression.java [hunk 3]
```diff
@@ -17,17 +17,18 @@
  */
 package org.wso2.ballerinalang.compiler.tree.expressions;
 
-import org.wso2.ballerinalang.compiler.semantics.model.symbols.BSymbol;
-
 /**
- * {@code BLangAccessibleExpression} represents an expression, where a {@code BLangAccessExpression} can be called
- * upon.
- * eg: list-construct, record literal, grouped expression, etc...
+ * {@code BLangValueExpression} represents a value access expression.
```
[the remark this answers is not in the document]
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangValueExpression.java]: fixed: d598ff7

**Review by rdhananjaya — APPROVED** [review]

**Pull request merged**